cmd/queue/main.go

```diff
@@ -201,8 +205,57 @@ func createAdminHandlers() *http.ServeMux {
 	return mux
 }
 
+func probeQueueHealthPath(port string) error {
+	url := fmt.Sprintf("http://127.0.0.1:%s%s", port, queue.RequestQueueHealthPath)
+
+	httpClient := &http.Client{
+		Transport: &http.Transport{
+			// Do not use the cached connection
+			DisableKeepAlives: true,
+		},
+		Timeout: 500 * time.Millisecond,
+	}
+
+	var lastErr error
+
+	err := wait.PollImmediate(100*time.Millisecond, time.Second, func() (bool, error) {
+		var res *http.Response
+		res, lastErr = httpClient.Get(url)
+
+		if res == nil {
+			return false, nil
+		}
+
+		defer res.Body.Close()
+		return res.StatusCode == http.StatusOK, nil
+	})
+
+	if lastErr != nil {
+		return fmt.Errorf("failed to probe: %s", lastErr)
+	}
+
+	// An http.StatusOK was never returned during probing
+	if err != nil {
+		return errors.New("probe returned not ready")
+	}
+
+	return nil
+}
+
 func main() {
 	flag.Parse()
+
+	if *probe {
+		port := strconv.Itoa(networking.QueueAdminPort)
+
+		if err := probeQueueHealthPath(port); err != nil {
+			fmt.Fprintln(os.Stderr, err)
```

Comment:
Should we use the logger here?

Reply:
I may be using the logger incorrectly, but experiments with it left me thinking we're better off with the simple print. When triggering an error and then calling `kubectl describe pods`...

...with fmt.Fprintln:
```
Events:
  Type     Reason     Age              From                                                  Message
  ----     ------     ----             ----                                                  -------
  Warning  Unhealthy  1s (x8 over 9s)  kubelet, my-cluster  Readiness probe failed: failed to probe: Get http://127.0.0.1:12312/health: dial tcp 127.0.0.1:12312: connect: connection refused    
```

...with logger.Error:
```
Events:
  Type     Reason     Age   From                                                  Message
  ----     ------     ----  ----                                                  -------
  Warning  Unhealthy  3s    kubelet, my-cluster  Readiness probe failed: {"level":"info","ts":"2019-05-27T18:24:25.406Z","caller":"logging/config.go:96","msg":"Successfully created the logger.","knative.dev/jsonconfig":"{\n  \"level\": \"info\",\n  \"development\": false,\n  \"outputPaths\": [\"stdout\"],\n  \"errorOutputPaths\": [\"stderr\"],\n  \"encoding\": \"json\",\n  \"encoderConfig\": {\n    \"timeKey\": \"ts\",\n    \"levelKey\": \"level\",\n    \"nameKey\": \"logger\",\n    \"callerKey\": \"caller\",\n    \"messageKey\": \"msg\",\n    \"stacktraceKey\": \"stacktrace\",\n    \"lineEnding\": \"\",\n    \"levelEncoder\": \"\",\n    \"timeEncoder\": \"iso8601\",\n    \"durationEncoder\": \"\",\n    \"callerEncoder\": \"\"\n  }\n}"}
{"level":"info","ts":"2019-05-27T18:24:25.406Z","caller":"logging/config.go:97","msg":"Logging level set to info"}
{"level":"warn","ts":"2019-05-27T18:24:25.406Z","caller":"logging/config.go:65","msg":"Fetch GitHub commit ID from kodata failed: \"ref: refs/heads/execprobe\" is not a valid GitHub commit ID"}                                                                                                                                                                 
...
~50 lines removed
...
{"level":"info","ts":"2019-05-27T18:24:25.406Z","logger":"queueproxy","caller":"util/env.go:43","msg":"USER_PORT=8080"}
{"level":"error","ts":"2019-05-27T18:24:26.407Z","logger":"queueproxy","caller":"queue/main.go:259","msg":"failed to probe: Get http://127.0.0.1:12312/health: dial tcp 127.0.0.1:12312: connect: connection refused","knative.dev/key":"default/configuration-example-mm4nx","knative.dev/pod":"configuration-example-mm4nx-deployment-686bf5d76c-mx878","stacktrace":"main.main\n\t/go/src/github.com/knative/serving/cmd/queue/main.go:259\nruntime.main\n\t/src/runtime/proc.go:200"}
```

Is there a way to make it less wordy?

Reply:
oh boy. Fine to leave as-is 😂